File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int solution(vector<string> board) {
    int n = board.size();
    int m = board[0].size();
    pair<int,int> goal;
    int map[101][101];
    int dy[4] = {1, 0, -1, 0};
    int dx[4] = {0, 1, 0, -1};
    
    queue<pair<int,int>> q;
    for(int i=0;i<n; i++) {
        for(int j=0;j<m; j++) {
            map[i][j] = 987654321;
            
            if(board[i][j] == 'R') {
                q.push({i, j});
                map[i][j] = 0;
            } else if(board[i][j] == 'G') {
                goal = {i, j};
            }
        }
    }
    
    while(!q.empty()) {
        int y = q.front().first;
        int x = q.front().second;
        q.pop();
        if(q.front() == goal) { continue; }
        
        for(int i=0;i<4;i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            
            if(ny>=0 && ny<n && nx >=0 && nx <m && board[ny][nx] != 'D') {
                while(true) {
                    if(ny>=0 && ny<n && nx >=0 && nx <m && board[ny][nx] != 'D') {
                        ny += dy[i];
                        nx += dx[i];
                    } else {
                        ny -= dy[i];
                        nx -= dx[i];
                        break;
                    }
                }
                
                if(map[ny][nx] > map[y][x] + 1) {
                    map[ny][nx] = map[y][x] + 1;
                    q.push({ny, nx});
                }
            }
        }
    }
    
    return map[goal.first][goal.second] == 987654321 ? -1 : map[goal.first][goal.second];
}

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.swift
import Foundation

var dx = [1,0,-1,0]
var dy = [0,1,0,-1]

func solution(_ board:[String]) -> Int {
    let board = board.map { $0.map { String($0) }}
    var queue: [(Int, Int)] = []
    var cntMap: [[Int]] = Array(repeating: Array(repeating: 987654321, count: board[0].count), count: board.count)
    
    for i in 0..<board.count {
        if !queue.isEmpty { break }
        for j in 0..<board[0].count {
            if board[i][j] == "R" {
                queue.append((i, j))
                cntMap[i][j] = 0
                break
            }
        }
    }
    
    while(!queue.isEmpty) {
        var position = queue.removeFirst()
        var y = position.0
        var x = position.1
        
        if board[y][x] == "G" {
            return cntMap[y][x]
        }
        
        for i in 0..<4 {
            var ny = y
            var nx = x
            
            while(true) {
                ny += dy[i]
                nx += dx[i]
                
                if ny < 0 || ny >= board.count || nx < 0 || nx >= board[0].count {
                    ny -= dy[i]
                    nx -= dx[i]
                    break
                }
                
                if board[ny][nx] == "D" {
                    ny -= dy[i]
                    nx -= dx[i]
                    break
                }
            }
            
            if cntMap[ny][nx] > cntMap[y][x] + 1 {
                cntMap[ny][nx] = cntMap[y][x] + 1
                queue.append((ny, nx))
            }
        }
    }
    
    return -1
}